src/lib.rs

```diff
@@ -34,9 +92,133 @@ impl fmt::Display for Priority {
+        Ok(IsNull::No)
+    }
+
+    fn accepts(ty: &Type) -> bool {
+        ty.name() == "priority"
+    }
+
+    tokio_postgres::types::to_sql_checked!();
+}
+
+impl FromStr for Priority {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "low" => Ok(Priority::Low),
+            "medium" => Ok(Priority::Medium),
+            "high" => Ok(Priority::High),
+            _ => Err(format!("Invalid priority: {}", s)),
+        }
+    }
+}
+
+impl<'a> FromSql<'a> for Priority {
+    fn from_sql(_ty: &Type, raw: &'a [u8]) -> Result<Self, Box<dyn Error + Send + Sync>> {
+        let s = std::str::from_utf8(raw)?;
+        s.parse().map_err(|e| {
+            Box::new(std::io::Error::new(std::io::ErrorKind::InvalidData, e))
+                as Box<dyn Error + Send + Sync>
+        })
+    }
+
+    fn accepts(ty: &Type) -> bool {
+        ty.name() == "priority"
+    }
+}
+
+#[derive(Debug, Clone)]
 pub struct Task {
     pub description: String,
     pub difficulty: Option<Difficulty>,
     pub priority: Option<Priority>,
     pub deadline: Option<NaiveDate>,
 }
+
+impl Task {
+    pub async fn insert_todo(&self) -> Result<(), Box<dyn Error>> {
+        let client = get_client().await?;
+
+        client
+            .execute(
+                "INSERT INTO todo (description, priority, difficulty, deadline) VALUES ($1, $2, $3, $4)",
+                &[&self.description, &self.priority, &self.difficulty, &self.deadline],
+            )
+            .await
+            .expect("Failed to insert task");
+
+        println!("Task \"{}\" created successfully!", self.description);
+
+        Ok(())
+    }
+
+    pub async fn delete_task(&self) -> Result<(), Box<dyn Error>> {
+        let client = get_client().await?;
+
+        client
+            .execute(
+                "DELETE FROM todo WHERE description = $1",
+                &[&self.description],
+            )
+            .await
+            .expect("Failed to delete task");
+
+        println!("Task \"{}\" deleted successfully!", self.description);
+
+        Ok(())
+    }
+
+    pub async fn list_tasks() -> Result<Vec<Task>, Box<dyn Error>> {
+        let client = get_client().await?;
+
+        let today = Local::now().date_naive();
+        let rows = client
+            .query(
+                "SELECT description, priority, difficulty, deadline FROM todo WHERE deadline = $1::date",
+                &[&today],
+            )
+            .await
+            .expect("Failed to fetch all tasks.");
+
+        let tasks: Vec<Task> = rows
+            .iter()
+            .map(|row| Task {
+                description: row.get(0),
+                priority: row.get(1),
+                difficulty: row.get(2),
+                deadline: row.get(3),
+            })
+            .collect();
+
+        Ok(tasks)
+    }
+
+    pub async fn update_task(&self, old_description: String) -> Result<(), Box<dyn Error>> {
```

Comment:
but we are listing the tasks to user and asking which one to edit, then `INSERT ... ON CONFLICT UPDATE` will not do as desired as the primary key is id of the entry. Either primary key should be description + deadline or it will always create a  new entry and never update

Reply:
> but we are listing the tasks to user and asking which one to edit

What does this mean? If we're listing tasks, wouldn't they all be objects of the `Task` struct anyway?

Reply:
But the task struct does not have id field and Task does not have `Display` trait implemented so we are just storing user's choice in `old_description`. We, currently, are not getting the object which needs to be changed. We are just getting the description of that task.

Reply:
i implemented `Display` trait for `Task`, now we can directly show the tasks in `inquire::Select()`. But still if we are not storing id in the struct itself, we have to update via old description, which is also not good because two tasks can have same name but different deadlines. `INSERT ... ON CONFLICT UPDATE` will also not work because primary key violation will never occur and a new entry will be created instead of updating existing query.

Reply:
> two tasks can have same description but different deadlines

Why? Description should be unique in itself

Reply:
So are we leaving that to the user?

Reply:
I mean, I can't imagine a scenario where a task has the exact same description with a different deadline

Reply:
In case of a detailed description it might not happen, but if the task is like 'Watch a movie', it can have 2 dates.  But yeah, if we are going with unique description then the code is alright.